### DISCUSSION TITLE
MeshGenerators during simulation?

### DISCUSSION THREAD

**Discussion by aeslaughter** [idaholab/moose#16635] [Q&A Meshing] [answered]
#16060 brings up a question. At what point should we invest time to enable the running a MeshGenerator after each timestep to allow dynamic mesh modification? For the problem in #16060 this would allow define/redefine the subdomains based on the level set? @friedmud @permcody @lindsayad @fdkong 

If there is desire and funding I would love to implement that, I have always thought it would be useful for this reason. There is a version of this in Mastodon that uses Aux variables that could go away as well.

**Comment by fdkong** [comment, marked as answer]
The mesh modification after each time step is already possible. Here is an example https://github.com/idaholab/moose/pull/16008

Are you asking for regenerating a different mesh  after each time step?

**Reply by aeslaughter** [reply]
I would like to modify the subdomains

**Comment by YaqiWang** [comment]
Will the entire mesh be regenerated or just applying some mesh generators on the existing mesh? This sounds to me should be done in a specific user object with `meshChanged` function. This may complicate our design a lot.

**Comment by hugary1995** [comment]
I think the phrase "dynamic mesh modification" is too general. There are mesh cutting, interface movement, element addition/deletion, dynamic remeshing, etc. How do you handle stateful properties in all scenarios in a unified way?

**Comment by ngrilli** [comment]
I am currently testing the approach in #16008 
to move elements from on subdomain to another.

This is useful for me to add elements for additive manufacturing simulations.
The material properties in the added elements are reinitialized properly, for instance plastic deformation

You can find some example tests here:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/ElementAddDelete

I am currently having some negative Jacobian error when I run large simulations with lots of element deleted at each time step
and I am investigating the issue. But I think it's not related to the method itself, it's my material model.

Best Regards,
Nicolò Grilli
National University of Singapore
